File: invoice-parse/libs/shared-py/invoice_shared/adapters/blob_store.py
"""Blob storage adapter — abstract interface + local filesystem implementation."""

from __future__ import annotations

import re
from abc import ABC, abstractmethod
from pathlib import Path

_UUID_RE = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", re.IGNORECASE)


class BlobStore(ABC):
    @abstractmethod
    def put(self, path: str, data: bytes) -> None: ...

    @abstractmethod
    def get(self, path: str) -> bytes: ...

    @abstractmethod
    def exists(self, path: str) -> bool: ...

    @abstractmethod
    def delete(self, path: str) -> None: ...


class LocalFsBlobStore(BlobStore):
    """Local filesystem blob store with path safety validation."""

    def __init__(self, base_path: str | Path) -> None:
        self._base = Path(base_path).resolve()
        self._base.mkdir(parents=True, exist_ok=True)

    def _safe_path(self, path: str) -> Path:
        if ".." in path:
            raise ValueError(f"Path traversal detected in blob path: {path}")

        # Validate UUID segments (tenant_id/job_id)
        parts = Path(path).parts
        # Strip leading separator
        clean_parts = [p for p in parts if p != "/"]
        if len(clean_parts) >= 2:
            tenant_id, job_id = clean_parts[0], clean_parts[1]
            if not _UUID_RE.match(tenant_id):
                raise ValueError(f"Invalid tenant_id in blob path: {tenant_id}")
            if not _UUID_RE.match(job_id):
                raise ValueError(f"Invalid job_id in blob path: {job_id}")

        resolved = (self._base / path.lstrip("/")).resolve()
        if not str(resolved).startswith(str(self._base)):
            raise ValueError(f"Path escapes base directory: {path}")
        return resolved

    def put(self, path: str, data: bytes) -> None:
        full = self._safe_path(path)
        full.parent.mkdir(parents=True, exist_ok=True)
        full.write_bytes(data)

    def get(self, path: str) -> bytes:
        return self._safe_path(path).read_bytes()

    def exists(self, path: str) -> bool:
        return self._safe_path(path).exists()

    def delete(self, path: str) -> None:
        p = self._safe_path(path)
        if p.exists():
            p.unlink()

File: invoice-parse/libs/shared-rs/src/adapters/blob_store.rs
use std::path::{Path, PathBuf};

use thiserror::Error;

#[derive(Debug, Error)]
pub enum BlobError {
    #[error("Path traversal detected: {0}")]
    PathTraversal(String),
    #[error("Invalid UUID segment in path: {0}")]
    InvalidUuid(String),
    #[error("Path escapes base directory: {0}")]
    PathEscape(String),
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),
}

/// Abstract blob storage interface.
pub trait BlobStore: Send + Sync {
    fn put(&self, path: &str, data: &[u8]) -> Result<(), BlobError>;
    fn get(&self, path: &str) -> Result<Vec<u8>, BlobError>;
    fn exists(&self, path: &str) -> Result<bool, BlobError>;
    fn delete(&self, path: &str) -> Result<(), BlobError>;
}

/// Local filesystem blob store with path safety validation.
pub struct LocalFsBlobStore {
    base: PathBuf,
}

impl LocalFsBlobStore {
    pub fn new(base_path: &str) -> Result<Self, BlobError> {
        let base = Path::new(base_path).canonicalize().or_else(|_| {
            std::fs::create_dir_all(base_path)?;
            Path::new(base_path).canonicalize()
        })?;
        Ok(Self { base })
    }

    fn safe_path(&self, path: &str) -> Result<PathBuf, BlobError> {
        if path.contains("..") {
            return Err(BlobError::PathTraversal(path.to_string()));
        }

        // Validate UUID segments (tenant_id/job_id)
        let clean = path.trim_start_matches('/');
        let parts: Vec<&str> = clean.split('/').collect();
        if parts.len() >= 2 {
            for segment in &parts[..2] {
                if uuid::Uuid::parse_str(segment).is_err() {
                    return Err(BlobError::InvalidUuid(segment.to_string()));
                }
            }
        }

        let full = self.base.join(clean);
        // Canonicalize parent to check prefix (file may not exist yet)
        let parent = full.parent().unwrap_or(&full);
        std::fs::create_dir_all(parent)?;
        let resolved_parent = parent.canonicalize()?;
        if !resolved_parent.starts_with(&self.base) {
            return Err(BlobError::PathEscape(path.to_string()));
        }
        Ok(full)
    }
}

impl BlobStore for LocalFsBlobStore {
    fn put(&self, path: &str, data: &[u8]) -> Result<(), BlobError> {
        let full = self.safe_path(path)?;
        std::fs::write(full, data)?;
        Ok(())
    }

    fn get(&self, path: &str) -> Result<Vec<u8>, BlobError> {
        let full = self.safe_path(path)?;
        Ok(std::fs::read(full)?)
    }

    fn exists(&self, path: &str) -> Result<bool, BlobError> {
        let full = self.safe_path(path)?;
        Ok(full.exists())
    }

    fn delete(&self, path: &str) -> Result<(), BlobError> {
        let full = self.safe_path(path)?;
        if full.exists() {
            std::fs::remove_file(full)?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    fn setup() -> (TempDir, LocalFsBlobStore) {
        let dir = TempDir::new().unwrap();
        let store = LocalFsBlobStore::new(dir.path().to_str().unwrap()).unwrap();
        (dir, store)
    }

    #[test]
    fn put_get_exists_delete_roundtrip() {
        let (_dir, store) = setup();
        let path = "a0000000-0000-0000-0000-000000000001/b0000000-0000-0000-0000-000000000002/input.pdf";
        let data = b"hello world";

        assert!(!store.exists(path).unwrap());
        store.put(path, data).unwrap();
        assert!(store.exists(path).unwrap());
        assert_eq!(store.get(path).unwrap(), data);
        store.delete(path).unwrap();
        assert!(!store.exists(path).unwrap());
    }

    #[test]
    fn rejects_path_traversal() {
        let (_dir, store) = setup();
        let result = store.put("../escape/file.txt", b"bad");
        assert!(matches!(result, Err(BlobError::PathTraversal(_))));
    }

    #[test]
    fn rejects_non_uuid_segments() {
        let (_dir, store) = setup();
        let result = store.put("not-a-uuid/also-bad/file.txt", b"data");
        assert!(matches!(result, Err(BlobError::InvalidUuid(_))));
    }
}
